src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                                MultiplyAdd(ref currentBuffer, TenPowMaxPartial, _partialValue);
+                                _partialValue = 0;
+                                _partialDigitCount = 0;
+                            }
+                        }
+
+                        // Check for nonzero digits after the decimal point.
+                        if (!endReached)
+                        {
+                            ReadOnlySpan<char> fracDigitsSpan = chunkDigits.Slice(intDigitsSpan.Length);
+                            for (int i = 0; i < fracDigitsSpan.Length; i++)
+                            {
+                                char digitChar = fracDigitsSpan[i];
+                                if (digitChar == '\0')
+                                {
+                                    break;
+                                }
+                                if (digitChar != '0')
+                                {
+                                    return false;
+                                }
+                            }
+                        }
+
+                        partialValue = _partialValue;
+                        partialDigitCount = _partialDigitCount;
+                        totalDigitCount = _totalDigitCount;
+
+                        return true;
+                    }
+                }
+                else
                 {
-                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    if (numberScale < 0)
                     {
                         result = default;
                         return false;
+
                     }
-                }
+                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
 
-                if (partialDigitCount > 0)
-                {
-                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+                    Span<uint> buffer = new uint[bufferSize];
+
+                    int bufferInd = buffer.Length - 1;
+                    uint currentBlock = 0;
+                    int shiftUntil = (totalDigitCount - 1) % MaxPartialDigits;
+                    int remainingIntDigitCount = totalDigitCount;
+                    foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                    {
+                        var digitsChunkSpan = digitsChunk.Span;
+                        ReadOnlySpan<char> intDigitsSpan = digitsChunkSpan.Slice(0, Math.Min(remainingIntDigitCount, digitsChunkSpan.Length));
+
+                        for (int i = 0; i < intDigitsSpan.Length; i++)
+                        {
+                            char digitChar = intDigitsSpan[i];
+                            Debug.Assert(char.IsDigit(digitChar));
+                            currentBlock = currentBlock * 10 + digitChar - '0';
```

Comment:
Just for clarity/readability:
```suggestion
                            currentBlock *= 10;
                            currentBlock += (digitChar - '0');
```